io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -146,6 +152,11 @@ data StmA s a where
   Retry        :: StmA s b
   OrElse       :: StmA s a -> StmA s a -> (a -> StmA s b) -> StmA s b
 
+  SayStm       :: String -> StmA s b -> StmA s b
+  OutputStm    :: Dynamic -> StmA s b -> StmA s b
```

Comment:
You only need these for debugging right? And `Debug.Trace` is not an option in those settings (I guess you want a chronological trace)?

Reply:
I use `SayStm` (via `say`) for debugging, yep. I basically provide it as an alternative to `debugTracer`.

I use `OutputStm` (via `traceSTM`) QuickCheck diagnostics (ie `tabulate` inputs and `Property`s that check event sequences) as well as for debugging.

Reply:
For future referents, we elaborated on the ticket; Issue #2725.